README.md
# Easy Tab & Accordion (ETA)

[![release](https://badgen.net/github/release/viivue/easy-tab-accordion/)](https://github.com/viivue/easy-tab-accordion/releases/latest)
[![minified](https://badgen.net/badge/minified/10KB/cyan)](https://www.jsdelivr.com/package/gh/viivue/easy-tab-accordion)
[![jsdelivr](https://data.jsdelivr.com/v1/package/gh/viivue/easy-tab-accordion/badge?style=rounded)](https://www.jsdelivr.com/package/gh/viivue/easy-tab-accordion)
[![license](https://badgen.net/github/license/viivue/easy-tab-accordion/)](https://github.com/viivue/easy-tab-accordion/blob/main/LICENSE)
[![Netlify Status](https://api.netlify.com/api/v1/badges/baaede11-84dc-46b3-b434-11aad923a156/deploy-status)](https://app.netlify.com/sites/easy-tab-accordion/deploys)

> Javascript library to create tabs or accordion.

## Getting started

### NPM Package

Install NPM package

```shell
npm i @viivue/easy-tab-accordion
```

Import

```js
import "@viivue/easy-tab-accordion";
```

### Download

👉 Self hosted - [Download the latest release](https://github.com/viivue/easy-tab-accordion/releases/latest)

```html

<script src="./easy-tab-accordion.min.js"></script>
```

👉 CDN Hosted - [jsDelivr](https://www.jsdelivr.com/package/gh/viivue/easy-tab-accordion)

```html
<!-- JS (10KB) -->
<script src="https://cdn.jsdelivr.net/gh/viivue/easy-tab-accordion@2.1.3/dist/easy-tab-accordion.min.js"></script>
```

## Initialize

To initialize an ETA script, always remember three things:

1. Wrapper
2. Trigger(s)
3. Receiver(s)

### With HTML

Using these HTML attributes to initialize without JavaScript.

```html
<!-- No Js init -->
<div data-eta>
    <!-- section 1 -->
    <div>
        <button data-eta-trigger="section-1">Section 1</button>
        <div data-eta-receiver="section-1">Content</div>
    </div>

    <!-- section 2 -->
    <div>
        <button data-eta-trigger="section-2">Section 2</button>
        <div data-eta-receiver="section-2">Content</div>
    </div>
</div>
```

Or using data-attributes (with JSON format) to initialize.

```html

<div data-eta='{ "id":"my-id", "animation":"slide", "hash":"false", "duration":"100"}'>
    <!-- section 1 -->
    <div>
        <button data-eta-trigger="section-1">Section 1</button>
        <div data-eta-receiver="section-1">Content</div>
    </div>

    <!-- section 2 -->
    <div>
        <button data-eta-trigger="section-2">Section 2</button>
        <div data-eta-receiver="section-2">Content</div>
    </div>
</div>
```

> ⚠️ Notice that value of `data-eta-trigger` and `data-eta-receiver` must be the same.

### With JavaScript

Assume that we have the HTML like below

```html
<!-- Custom HTML -->
<div class="my-accordion">
    <!-- section 1 -->
    <div>
        <button data-trigger="section-1">Section 1</button>
        <div class="content" id="section-1">Content</div>
    </div>

    <!-- section 2 -->
    <div>
        <button data-trigger="section-2">Section 2</button>
        <div class="content" id="section-2">Content</div>
    </div>
</div>
```

```js
// Init
ETA.init({
    el: document.querySelector('.my-accordion'), // DOM element
    trigger: '[data-trigger]', // CSS selector
    triggerAttr: 'data-trigger', // attribute name
    receiver: '.content', // CSS selector
    receiverAttr: 'id', // attribute name
});
```

## Options

### Selectors

| Name         | Type        | Default               | Required | Description                                |
|--------------|-------------|-----------------------|----------|--------------------------------------------|
| el           | DOM element | `[data-eta]`          | ✅        | Wrapper element                            |
| trigger      | string      | `[data-eta-trigger]`  | ✅        | CSS selector for trigger elements          |
| triggerAttr  | string      | `data-eta-trigger`    | ✅        | Attribute name of trigger elements         |
| receiver     | string      | `[data-eta-receiver]` | ✅        | CSS selector for receiver elements         |
| receiverAttr | string      | `data-eta-receiver`   | ✅        | Attribute name of receiver elements        |
| activeClass  | string      | `"active"`            | ❌        | Class name for active trigger and receiver |

> ⚠️ Notice that value of `triggerAttr` and `receiverAttr` must be the same.

### Animation

| Name      | Type   | Default   | Description                                                                                               |
|-----------|--------|-----------|-----------------------------------------------------------------------------------------------------------|
| animation | string | `"slide"` | `"slide"` for accordion style (slide up and slide down), `"fade"` for tabbed style (fade in and fade out) |
| duration  | number | `450`     | Duration of animation in milisecond                                                                       |

### Hash

| Name       | Type    | Default | Description                                         |
|------------|---------|---------|-----------------------------------------------------|
| hash       | boolean | `false` | Update hash URL                                     |
| hashScroll | boolean | `false` | Scroll into view when page loaded with a valid hash |

### Responsive

| Name           | Type  | Default | Description             |
|----------------|-------|---------|-------------------------|
| liveBreakpoint | array | `[]`    | An array of two numbers |

`liveBreakpoint` defines a range to enable ETA, when the browser's width is outside this range ETA will be destroyed (
detecting via window resizing event).

For instance:

- `liveBreakpoint:[99999,768]`: destroy ETA on screen that smaller than 768px
- `liveBreakpoint:[1024,-1]`: destroy ETA on screen that bigger than 1024px

### Open and close

| Name          | Type         | Default | Description                                                                                  |
|---------------|--------------|---------|----------------------------------------------------------------------------------------------|
| activeSection | number/array | `0`     | Index(s) of section to be active on init, array input only available for `animation:"slide"` |

#### For `animation:"slide"` only

| Name             | Type    | Default | Description                                     |
|------------------|---------|---------|-------------------------------------------------|
| allowCollapseAll | boolean | `false` | Allow to collapse all sections at the same time |
| allowExpandAll   | boolean | `false` | Allow to expand all sections at the same time   |

### HTML attributes

Add these attributes on the wrapper element.

| Attribute                   | As for option       | 
|-----------------------------|---------------------|
| `data-eta-animation="fade"` | `animation: "fade"` |
| `data-eta-hash`             | `hash: true`        | 
| `data-eta-hash-scroll`      | `hashScroll: true`  |

## Events

| Name                             | Description | 
|----------------------------------|-------------|
| `onBeforeInit: (data) => {}`     |             |
| `onAfterInit: (data) => {}`      |             |
| `onBeforeOpen: (data,el) => {}`  |             |
| `onBeforeClose: (data,el) => {}` |             |
| `onAfterOpen: (data,el) => {}`   |             |
| `onAfterClose: (data,el) => {}`  |             |
| `onDestroy: (data) => {}`        |             |
| `onUpdate: (data) => {}`         |             |

## Methods

| Name            | Usage                      | Description                 | 
|-----------------|----------------------------|-----------------------------|
| `toggle`        | `eta.toggle(id)`           | Toggle a section by ID      |
| `toggleByIndex` | `eta.toggleByIndex(index)` | Toggle a section by index   |
| `destroy`       | `eta.destroy()`            | Remove all style and events |
| `init`          | `eta.init()`               | Could be use after destroy  |
| `update`        | `eta.update()`             | Update styling              |

Get the instance with JS init

```js
ETA.init({
    id: 'my-eta'
});

const eta = ETA.get('my-eta');

// use methods
eta.update();
```

## Deployment

```shell
# Start dev server
npm run dev

# Distribute production files (set new version in `package.json` first)
npm run prod

# Build dev site (for Netlify only)
npm run build
```

## License

[MIT License](https://github.com/viivue/easy-tab-accordion/blob/main/LICENSE)

Copyright (c) 2023 ViiVue

CHANGELOG.md
# Changelog

### [2.1.3] - 2023-01-13

- Improve performances

### [2.1.2] - 2022-12-21

- Slide down function > fix a minor bug
- Fade animation missing position absolute
- Init ETA by data-attribute with JSON format

### [2.1.1] - 2022-12-09

- Minor bug of `slideDown` when initial

### [2.0.0] - 2022-07-03

- Refactor developing environment using webpack with ES6 modules
- Allow to expand all with `allowExpandAll`
- Method `init` and `update`
- Add prettier demos

### [1.0.1] - 2022-06-23

- Settings via attributes
- Allow collapse all

### [1.0.0] - 2022-04-04

- Accordion and tabs.
- Fade and slide effect.
- Destroy event and method.
- Hash support.
- Breakpoint support.

src/_index.js
import {slideDown, slideUp, destroySlide, updateSlide} from "./slide";
import {fadeIn, fadeOut, destroyFade, updateFade} from "./fade";
import {getHash, isValidHash, hashScroll, updateURL} from "./hash";
import {
    validID,
    getToggleState,
    getIndexById,
    getElements,
    removeActiveClass, addActiveClass, getIdByIndex, defaultActiveSections, log, getOptions
} from "./helpers";
import {debounce, uniqueId} from "./utils";
import {initSetup, onLoad, onResize} from "./methods";
import {isLive, validBreakpoints} from "./responsive";

export class EasyTabAccordion{
    constructor(options){
        this._class = {
            enabled: 'easy-tab-accordion-enabled',
            active: 'active'
        };
        this._attr = {
            container: 'data-eta',
            trigger: 'data-eta-trigger',
            receiver: 'data-eta-receiver',
            hash: 'data-eta-hash',
            hashScroll: 'data-eta-hash-scroll',
            animation: 'data-eta-animation'
        };
        this.defaultOptions = {
            // selectors
            el: document.querySelector(`[${this._attr.container}]`), // DOM element
            id: uniqueId('eta-'),
            trigger: `[${this._attr.trigger}]`, // string selector
            triggerAttr: this._attr.trigger, // attribute name
            receiver: `[${this._attr.receiver}]`, // string selector
            receiverAttr: this._attr.receiver, // attribute name
            activeClass: this._class.active,

            // animation
            animation: 'slide', // slide, fade
            duration: 450,

            // hash
            hash: false, // update hash URL
            hashScroll: false, // scroll into view when page loaded with a valid hash

            // responsive
            liveBreakpoint: [], // [1920, 1024] => destroy if window.width if bigger than 1920 or less than 1024

            // avoid double click
            avoidDoubleClick: true,

            // dev mode => enable console.log
            dev: false,

            // open/close
            activeSection: 0, // default opening sections, will be ignored if there's a valid hash, allow array of index [0,1,2] for slide animation only
            allowCollapseAll: false, // for slide animation only
            allowExpandAll: false, // for slide animation only

            // events
            onBeforeInit: (data) => {
            },
            onAfterInit: (data) => {
            },
            onBeforeOpen: (data, el) => {
            },
            onBeforeClose: (data, el) => {
            },
            onAfterOpen: (data, el) => {
            },
            onAfterClose: (data, el) => {
            },
            onDestroy: (data) => {
            },
            onUpdate: (data) => {
            },
        };

        // save options
        this.originalOptions = options;
        // init
        this.init();

        // avoid double click
        this.isAnimating = false;
    }

    init(){
        // setup
        this.options = {...this.defaultOptions, ...this.originalOptions};

        if(!this.options.el){
            log(this, 'warn', 'ETA Error, target not found!');
            return;
        }

        this.wrapper = this.options.el;
        this.id = '';
        this.current_id = '';
        this.previous_id = '';
        this.type = '';
        this.hasInitialized = false;
        this.enabled = validBreakpoints(this) ? isLive(this) : true;
        this.count = this.wrapper.querySelectorAll(this.options.trigger).length;

        // check if ETA has already initialized
        if(this.wrapper.classList.contains(this._class.enabled)){
            log(this, 'ETA has initialized');
            return;
        }

        // check the condition at openPanel, when calls close others (because there is no active element at begin)
        this.isFirst = true;

        // update hash from attribute
        this.options.hash = this.wrapper.hasAttribute(this._attr.hash) === true ? true : this.options.hash;
        this.options.hashScroll = this.wrapper.hasAttribute(this._attr.hashScroll) === true ? true : this.options.hashScroll;
        // update animation from attribute
        const animationValue = this.wrapper.getAttribute(this._attr.animation);
        this.options.animation = animationValue !== null ? animationValue : this.options.animation;

        // get options init by data attribute (JSON format)
        this.options = getOptions(this);

        // assign id to wrapper
        this.wrapper.setAttribute(this._attr.container, this.id);

        if(this.count < 1){
            log(this, 'warn', 'Quit init due to child panels not found', this);
            return;
        }

        // init
        if(this.enabled && !this.hasInitialized) initSetup(this);
        if(!this.enabled && this.hasInitialized) this.destroy();

        // toggle via hash
        if(this.enabled){
            if(isValidHash(this)){
                this.toggle(getHash().id, 'hash');
            }else{
                defaultActiveSections(this);
            }
        }

        // watch for resize/load events
        window.addEventListener('resize', debounce(e => onResize(this, e), 300));
        window.addEventListener('load', e => onLoad(this, e));
    }

    destroy(){
        this.hasInitialized = false;
        this.wrapper.classList.remove(this._class.enabled);

        // loop through triggers
        this.wrapper.querySelectorAll(this.options.trigger).forEach(trigger => {
            trigger.outerHTML = `${trigger.outerHTML}`; // a trick to remove all events by cloning itself
        });

        // loop through receivers
        this.dataset = [];

        // reset CSS for fade animation
        switch(this.options.animation){
            case "slide":
                destroySlide(this);
                break;
            case "fade":
                destroyFade(this);
                break;
        }

        // event: onDestroy
        this.options.onDestroy(this);
    }

    update(){
        switch(this.options.animation){
            case "slide":
                updateSlide(this);
                break;
            case "fade":
                updateFade(this);
                break;
        }

        // event: onUpdate
        this.options.onUpdate(this);
    }

    openPanel(id = this.current_id){
        if(!validID(this, id)) return;

        const beforeOpen = () => {
            // update section status
            this.dataset[getIndexById(this, id)].active = true;

            // update URL
            updateURL(this, id);

            // events
            this.options.onBeforeOpen(this);

            // start animation
            this.isAnimating = true;
            log(this, 'log', 'Start animation.');

            // stop animation
            clearTimeout(this.animationTimeout);
            this.animationTimeout = setTimeout(() => {
                // toggle animating status
                this.isAnimating = false;
                log(this, 'log', 'Stop animation.');
            }, this.options.duration + 10);
        };

        // event: on Before Open
        beforeOpen();

        // event: on After Open
        const afterOpen = (target) => {
            hashScroll(this);

            this.options.onAfterOpen(this, target);

            // log
            log(this, 'log', 'after open', id);
        };

        // open
        const {current} = getElements(this, id);
        switch(this.options.animation){
            case 'slide':
                current.forEach(el => slideDown(el, this.options.duration, () => afterOpen(el)));
                break;
            case 'fade':
                current.forEach(el => fadeIn(el, this.options.duration, () => afterOpen(el)));
                break;
        }

        // update classes
        addActiveClass(this, id);

        // close all others
        const closeAllOthers = this.options.animation === 'fade' || this.options.animation === 'slide' && !this.options.allowExpandAll;
        if(closeAllOthers) this.dataset.filter(x => x.id !== id).forEach(item => {
            if(item.active || this.isFirst){
                this.closePanel(item.id);
            }
        });
        if(this.isFirst){
            current[0].style.display = 'block';
            this.isFirst = false;
        }
    }

    closePanel(id = this.current_id){
        if(!validID(this, id)) return;

        // event: on Before Close
        this.options.onBeforeClose(this);

        // event: on After Close
        this.dataset[getIndexById(this, id)].active = false;
        const afterClose = (target) => {
            this.options.onAfterClose(this, target);

            // log
            log(this, 'log', 'after close', id);
        };

        // close animation
        const {current} = getElements(this, id);
        switch(this.options.animation){
            case 'slide':
                current.forEach(el => slideUp(el, this.options.duration, () => afterClose(el)));
                break;
            case 'fade':
                current.forEach(el => fadeOut(el, this.options.duration, () => afterClose(el)));
                break;
        }

        // update classes
        removeActiveClass(this, id);
    }

    toggle(id, type = 'undefined', force = false){
        // exit if id is not found
        if(!validID(this, id)){
            log(this, 'warn', 'invalid id');
            return;
        }

        const toggleState = getToggleState(this, id);
        if(toggleState === 0) return;

        // update data
        this.type = type;
        this.previous_id = this.current_id ? this.current_id : getIdByIndex(this, 0);
        if(toggleState === 1){
            // open
            this.current_id = id;
            this.openPanel(this.current_id);
        }else{
            // close
            this.closePanel(id);
        }
    }

    toggleByIndex(index){
        this.toggle(getIdByIndex(this, index));
    }
}

/**
 * Private class Controller
 * This class will hold instances of the library's objects
 */
class Controller{
    constructor(){
        this.instances = [];
    }

    add(instance){
        this.instances.push(instance);
    }

    get(id){
        return this.instances.filter(instance => instance.id === id)[0];
    }
}


/**
 * Public library data
 * access via window.ETAController
 */
window.ETAController = new Controller();

/**
 * Public library object
 * access via window.ETA
 */
window.ETA = {
    // init new instances
    init: (options = undefined) => {
        // no param
        if(typeof options === 'undefined'){
            // init with attribute
            document.querySelectorAll('[data-eta]').forEach(el => {
                window.ETAController.add(new EasyTabAccordion({el, ...options}));
            });
        }

        window.ETAController.add(new EasyTabAccordion(options));
    },
    // Get instance object by ID
    get: id => window.ETAController.get(id)
};

window.ETA.init();